pkg/resourcemanager/mysql/mysqlhelper.go

```diff
@@ -88,30 +93,112 @@ func ConnectToSQLDBAsCurrentUser(
 	return db, err
 }
 
+// parseGrantsRow returns the grants and the DB for the row, or an error
+func parseGrantsRow(row string) (map[string]struct{}, string, error) {
+	matches := grantsRegex.FindStringSubmatch(row)
+	if len(matches) < 6 {
+		return nil, "", errors.Errorf("grants row %q didn't match expected regex", row)
+	}
+
+	grants := strings.Split(matches[1], ", ")
+	db := matches[2]
+
+	// Force lowercase here -- assumption is DBs don't differ by case
+	db = strings.ToLower(db)
+
+	// Put the grants in a map since we don't care about ordering
+	grantsMap := make(map[string]struct{})
+	for _, grant := range grants {
+		grantsMap[grant] = struct{}{}
+	}
+
+	return grantsMap, db, nil
+}
+
+// ExtractUserRoles extracts the roles the user has. The result is a map of database to roles
+func ExtractUserRoles(ctx context.Context, user string, db *sql.DB) (map[string]map[string]struct{}, error) {
+	if err := helpers.FindBadChars(user); err != nil {
+		return nil, errors.Wrapf(err, "problem found with username")
+	}
+
+	// TODO: Possibly need to fully qualify this with database name?
+	rows, err := db.QueryContext(ctx, "SHOW GRANTS FOR ?", user)
```

Comment:
The output of SHOW GRANTS is really intended for humans, can we get the information we need by querying the information schema? That would hopefully give us individual rows for each role so you wouldn't need the nasty parsing. (Trying to find a clear reference for doing this - I'm more familiar with the postgres ones for this kind of info.)

Reply:
I just tried it out, I think selecting from `information_schema.user_privileges` would get what you want, since we're granting on everything?

Reply:
Trying this out

Reply:
This works. I've changed to this which is quite a lot cleaner and makes me feel significantly better about the chances this is going to work in practice with real users and permissions. I left a note stating that this works only because we're limited in what permissions we manage (only per-DB, rather than per-table or per-column).